Fix: Return 405 for non-GET/HEAD requests to pages in dev mode

Rezakarimzadeh98 opened the pull request:
## Description

Fixes #38863

This PR ensures that `next dev` returns `405 Method Not Allowed` for non-GET/HEAD HTTP requests to page routes, making the behavior consistent with `next start`.

## Changes

- Added HTTP method validation in `router-server.ts` for page routes
- Non-GET/HEAD requests (POST, PUT, DELETE, PATCH, etc.) to pages now return 405
- API routes continue to accept all HTTP methods
- Added `Allow: GET, HEAD` header in 405 responses
- Added comprehensive integration tests

## Before

In development mode, POST requests to pages returned `200 OK`:
```bash
curl -X POST http://localhost:3000/ 
# Returns 200 OK (incorrect)

Commit:
Fix: Return 405 for non-GET/HEAD requests to pages in dev mode

- Added logic in router-server.ts to check HTTP method for page routes
- Non-GET/HEAD requests to pages now return 405 Method Not Allowed
- API routes continue to accept all HTTP methods
- Added comprehensive integration tests

Fixes #38863

ijjk commented:
Allow CI Workflow Run

- [ ] approve CI run for commit: 9c78071355ed34dd152920263f88fe8802d0a52e

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

Commit:
fix: resolve TypeScript errors in test mocks

Rezakarimzadeh98 committed:
Merge branch 'canary' into fix/issue-38863-post-request-405

Rezakarimzadeh98 committed:
Merge branch 'canary' into fix/issue-38863-post-request-405

Rezakarimzadeh98 committed:
fix: remove incorrectly merged test file

This test file was meant for PR #86408 (pageExtensions feature) but was
incorrectly merged into this branch via canary merge. Removing it since
getPageExtensions doesn't exist in this branch.

Fixes TypeScript error in test suite.

Rezakarimzadeh98 committed:
Merge branch 'canary' into fix/issue-38863-post-request-405

Rezakarimzadeh98 committed:
Merge branch 'canary' into fix/issue-38863-post-request-405

Rezakarimzadeh98 committed:
Merge branch 'canary' into fix/issue-38863-post-request-405

Rezakarimzadeh98 committed:
test: shorten isolated install dir names on Windows

Rezakarimzadeh98 committed:
Merge remote-tracking branch 'upstream/canary' into fix/issue-38863-post-request-405

Rezakarimzadeh98 committed:
chore: drop unrelated changes

Rezakarimzadeh98 committed:
Merge branch 'canary' into fix/issue-38863-post-request-405

Rezakarimzadeh98 commented:
@ijjk Ready for CI approval! 🚀

All review feedback has been addressed and the code is ready for testing. Looking forward to getting this merged!

Rezakarimzadeh98 committed:
test: format create-next-install

Rezakarimzadeh98 committed:
Merge branch 'canary' into fix/issue-38863-post-request-405

Rezakarimzadeh98 commented:
Hi @ijjk! 👋

The HTTP methods test suite has been validated locally and all 12 tests pass successfully. The CI failures appear to be related to self-hosted runner infrastructure (the "job started hook" error mentioning `/root/validate-job.sh` suggests a runner permission/setup issue rather than a code problem).

Could you please approve/rerun CI for the latest commit (da487d63a5feaa4fb94e86e3b4ce2d7efcd651c6)?

Thank you! 🙏

Rezakarimzadeh98 committed:
Merge branch 'canary' into fix/issue-38863-post-request-405

Rezakarimzadeh98 committed:
Merge branch 'canary' into fix/issue-38863-post-request-405

Rezakarimzadeh98 committed:
Merge branch 'canary' into fix/issue-38863-post-request-405

Rezakarimzadeh98 committed:
Merge branch 'canary' into fix/issue-38863-post-request-405
